### PR TITLE
Implement `passwordLength` field for service hosting generated accounts

### PR DESCRIPTION
**Pull request by wolveix**
Hey guys. First off, I implemented a working `testConnection` function for DirectAdmin, as the previous implementation did not work.

Secondly, I ran into an issue where my DirectAdmin installation has a minimum password length of 12 characters, while the FOSSBilling password generation only generates a string up to 10-characters long. I thought about a few different implementations, and ultimately landed on implementing a generic `passwordLength` field, usable by all of the service hosts.

I've made the change everywhere it seemed necessary (including the SQL migrations); however, it's clear that I've missed something as the value does not get submitted to the database, and therefore does not get retrieved either (regardless of it being done in the creation/or update phase).

I'd appreciate an overall review of this PR. Any suggestions/recommendations on a better implementation would be greatly appreciated! I haven't written PHP in approximately 5 years, so apologies if there's something obvious I've missed.

### CODE REVIEW

**Comment by wolveix** [comment]
@BelleNottelling

I've left it as a draft as the `passwordLength` field doesn't currently get properly stored in the database, which I'm sure is something simple I'm overlooking. 

To clarify my stance on alternative implementations, DirectAdmin allows you to configure a minimum and maximum password length. Similarly, you can do the same for usernames (which I suspect will be my next issue). This is why I ultimately landed on this implementation, as no default values would work. The only alternative in my eyes would be to note the maximum length returned by the API on error, regenerate the password at that stage, and then re-attempt the request. That feels incredibly messy to me, though.

File: src/library/Server/Manager/Directadmin.php [hunk 1]
```diff
@@ -315,152 +394,176 @@ public function modifyAccount(Server_Account $a)
...
 
-        $r = $this->_parseResponse($data);
+        $response = $this->_parseResponse($data);
 
-        if(isset($r['error']) && $r['error'] == 1) {
-            $placeholders = [':action:' => $command, ':type:' => 'DirectAdmin'];
-            throw new Server_Exception('Failed to :action: on the :type: server, check the error logs for further details', $placeholders);
+        if(isset($response['error']) && $response['error'] == 1) {
+            $placeholders = [':action:' => $command, ':type:' => 'DirectAdmin', ':error:' => $response['text'] . ': ' . $response['details']];
+            throw new Server_Exception('Failed to :action: on the :type: server: :error:', $placeholders);
```
Comment (BelleNottelling) [inline src/library/Server/Manager/Directadmin.php:534]: I'm typically not a big fan on sending errors to the front-end like this, clients will also get shown this message and sometimes returned errors can be a bit too.. *showy* if that makes sense.

**Comment by wolveix** [comment]
I've just noticed an edge case in the existing DirectAdmin implementation. As it replaces the stored user account's password with asterisks, uncancelling an order results in the password being set on DirectAdmin as such. It looks like WHM and Plesk have different implementations, which I'm looking over now. Otherwise, I'll just re-gen the password on uncancel

File: src/modules/Servicehosting/html_admin/mod_servicehosting_index.html.twig [hunk 2]
```diff
@@ -220,7 +220,13 @@
                         </div>
                     </div>
                     <div class="mb-3 row">
-                        <label class="form-label col-3 col-form-label">{{ 'Use Secure connection'|trans }}:</label>
+                        <label class="form-label col-3 col-form-label">{{ 'Password length'|trans }}:</label>
+                        <div class="col">
+                            <input class="form-control" type="text" name="passwordlength" value="" placeholder="{{ 'Generated password length. Use blank to use default (10). Used when creating user accounts'|trans }}">
```
Comment (BelleNottelling) [inline src/modules/Servicehosting/html_admin/mod_servicehosting_index.html.twig:225]: I'm not a fan of this wording. Perhaps something like this instead:
Length of generated passwords when creating user accounts. Defaults to 10.
Or:
Generated password length. Defaults to 10 when blank. Used when creating user accounts.

File: src/modules/Servicehosting/html_admin/mod_servicehosting_server.html.twig [hunk 3]
```diff
@@ -87,7 +87,13 @@
                 </div>
             </div>
             <div class="mb-3 row">
-                <label class="form-label col-3 col-form-label">{{ 'Use Secure connection'|trans }}:</label>
+                <label class="form-label col-3 col-form-label">{{ 'Password length'|trans }}:</label>
+                <div class="col">
+                    <input class="form-control" type="text" name="passwordlength" value="{{ server.passwordlength }}" placeholder="{{ 'Generated password length. Use blank to use default (10). Used when creating user accounts'|trans }}">
```
Comment (BelleNottelling) [inline src/modules/Servicehosting/html_admin/mod_servicehosting_server.html.twig:92]: Same here regarding the wording

File: src/library/Server/Manager/Directadmin.php [hunk 4]
```diff
@@ -315,152 +394,176 @@ public function modifyAccount(Server_Account $a)
...
-            if($post) {
-                $request = $httpClient->request('POST', $url, [
-                    'body'  => $fields,
-                ]);
-            } else {
-                $request = $httpClient->request('GET', $url);
-            }
-
+            // If it's a POST request, include the fields in the request body
+            $request = $http_client->request($post ? 'POST' : 'GET', $url, $post ? ['body' => $fields] : []);
```
Comment (BelleNottelling) [inline src/library/Server/Manager/Directadmin.php:513]: I personally think this change makes the code far less readable and would prefer it if the original code was kept.

**Comment by wolveix** [comment]
TL;DR for the above:
- Fixed issue with all service host adapters where upon an order being uncancelled, the account's password would be remotely set as asterisks
- Implemented DirectAdmin connection testing that works
- Implemented DirectAdmin package select (you can manually specify a custom `package` value, which will now be checked against the DirectAdmin server)
- Implemented `passwordlength` configurable value for all service host adapters

**Review by BelleNottelling — COMMENTED** [review]
Some remarks:
1. last_patch needs to be updated whenever adding new patches: https://github.com/FOSSBilling/FOSSBilling/blob/main/src/install/sql/content.sql#L318. If we miss this step, updating breaks for existing installs as it'll try to perform a task that's already been done on the DB.
2. `passwordlength` needs to be added to the cspell ignore list: https://github.com/FOSSBilling/FOSSBilling/blob/main/cspell.json#L42.
3. It's not really specific to this PR and is just generally how things are setup right now, but I think we need to stop doing separate DB fields for these configs and instead move to using JSON in a single column. It's more flexible for server managers and would allow us to continue what was started in #544 to give the server managers themselves far more control over the config options that are provided.
4. A couple of comments I put on the code itself.
5. Finally, I just wanted to appreciate all the added comments to document the code. It's a nice touch

File: src/library/Server/Manager/Directadmin.php (hunk 1 above)
Comment (wolveix) [reply src/library/Server/Manager/Directadmin.php:534]: Completely agree, apologies for the oversight! I'll adjust this, and log the error internally

File: src/library/Server/Manager/Directadmin.php (hunk 4 above)
Comment (wolveix) [reply src/library/Server/Manager/Directadmin.php:513]: Sure thing, no problem at all

**Comment by BelleNottelling** [comment]
> Fixed issue with all service host adapters where upon an order being uncancelled, the account's password would be remotely set as asterisks

That's a good find & fix, thank you.
I'm honestly not sure why we haven't deleted that DB row entirely yet, but that can be done in a separate PR

**Comment by wolveix** [comment]
Thanks for the fast review! I really appreciate it :)

I believe I've satisfied your comments, though please do double check! I completely agree on moving to a field containing JSON approach, and I'd be happy to help with that as/when needed.

As a minor sidenote, for context I would have preferred naming the field `passwordLength` instead, but the only other field on the model containing multiple words (`accesshash`) didn't follow camelcase, so 🤷 I noticed that there's a real mix of camel and snake case throughout the codebase. I did change a few things here and there, and modified a few single letter variable names to be more descriptive. I'd love to know (and apologies if you've already noted it elsewhere) which naming convention the codebase _should_ be using. 

As I integrate FOSSBilling more tightly into our infrastructure, I suspect I'll be spending more time helping out. Thanks again!

**Comment by wolveix** [comment]
I'm confused about the CI tests failing, as I can't see an obvious issue within this PR. Again, my apologies for my lack of familiarity with PHP, it's been a while since I've worked with it like this. Any insight would be greatly appreciated!

**Comment by BelleNottelling** [comment]
> I noticed that there's a real mix of camel and snake case throughout the codebase

Indeed, you can thank BoxBilling for that. For new code / heavy rewrites I'd definitely go for camelCase, however in cases where you're updating code I think it would be fine to do either camelCase or follow what's already in the surrounding code.



> I'm confused about the CI tests failing

Pulled the branch. It looks like this is a great example of the many poorly written tests from the BoxBilling project.
Many of them either don't test much, are implemented wrong, or I've even seen cases of a test that required broken behavior to pass.

In this case, it's a test that's not correctly implemented and it's not mocking all of the required data for it to pass.
I typically have been removing these tests as they've proven to be largely unhelpful. Feel free to remove it or to correct the testcase itself if you want, either one is fine.

The longer-term goal is to actually deprecate and eventually remove these tests entirely and instead perform tests via live docker instances where there's an actual config deployed with a real database as the current ones simply mock **everything** and are nearly useless for actually finding problems.

**Comment by wolveix** [comment]
Hah, go figure! In which case, I may rework this and switch over to camelCase. Might as well get it right the first time for new code! I may spend some time tonight/tomorrow doing so, as well as reworking more of the DirectAdmin implementation.

I'll also apologize for doing many different things within one PR. I tend to do that when first contributing to a project, since I end up stumbling across 101 things that _could_ be improved (such as docs). I'm a big proponent of documenting and testing everything (particularly given that I work with Go on a daily basis, and love the language).

Once again I appreciate your fast reviews and responses! I'll get those changes done tonight/tomorrow, and hopefully you and another person will be able to review them soon :) Thanks!

**Comment by wolveix** [comment]
@BelleNottelling sorry to tag! Were you waiting on me to do anything else?

**Comment by BelleNottelling** [comment]
Oh, @wolveix 
I honestly wasn't sure if it was ready for review or not, but if it is I'll give it a review now and let you know if I find anything

File: src/library/Server/Manager/Directadmin.php [hunk 5]
```diff
@@ -52,135 +74,192 @@ public static function getForm()
...
     }
 
-    public function _getPort()
+    /**
+     * Returns the port number for the DirectAdmin server.
+     * If the port is not set in the configuration, it defaults to '2222'.
+     *
+     * @return float|int|string The port number.
+     */
+    public function _getPort(): float|int|string
```
Comment (BelleNottelling) [inline src/library/Server/Manager/Directadmin.php:83]: Why a float? A port can't be a float

File: src/library/Server/Manager/Directadmin.php [hunk 6]
```diff
@@ -424,43 +521,56 @@ private function _request($command, $fields = array(), $post = true)
...
-            $e = new Server_Exception('HttpClientException: :error', [':error' => $error->getMessage()]);
-            $this->getLog()->err($e);
-            throw $e;
+            $exception = new Server_Exception('HttpClientException: :error', [':error' => $error->getMessage()]);
+            $this->getLog()->err($exception);
+            throw $exception;
         }
 
-        if(strlen(strstr($data, 'DirectAdmin Login')) > 0) {
+        if(strlen(strstr($data, '<!doctype html>')) > 0 || strlen(strstr($data, 'DirectAdmin Login')) > 0) {
```
Comment (BelleNottelling) [inline src/library/Server/Manager/Directadmin.php:529]: I think we probably should have a comment here explaining why this logic is used.
It's not particularly clear

File: src/library/Server/Manager/Directadmin.php [hunk 7]
```diff
@@ -424,43 +521,56 @@ private function _request($command, $fields = array(), $post = true)
...
 
-        $r = $this->_parseResponse($data);
+        $response = $this->_parseResponse($data);
 
-        if(isset($r['error']) && $r['error'] == 1) {
-            $placeholders = [':action:' => $command, ':type:' => 'DirectAdmin'];
-            throw new Server_Exception('Failed to :action: on the :type: server, check the error logs for further details', $placeholders);
+        if(isset($response['error']) && $response['error'] == 1) {
+            $this->getLog()->err('Failed to ' . $command . ' on the DirectAdmin server: ' . $response['text'] . ': ' . $response['details']);
+            throw new Server_Exception('Failed to complete the operation on the DirectAdmin server');
```
Comment (BelleNottelling) [inline src/library/Server/Manager/Directadmin.php:541]: Any reason not to use the original message?
All of these are turned into translation strings so it's preferable to not create a bunch of unique ones unless there's a specific reason

**Review by BelleNottelling — COMMENTED** [review]
Just a few very minor things I noticed, otherwise this looks pretty good

**Comment by jaapmarcus** [comment]
If possible I would prefer where we store the password length in a text while json encode. Instead storing it it the table directly as it would allow plugin developers to expend other server plugins where it is needed any way... 

It would solve a lot of duplicated work in the future ...

File: src/library/Server/Manager/Directadmin.php (hunk 6 above)
Comment (wolveix) [reply src/library/Server/Manager/Directadmin.php:529]: Ah, I did add a comment somewhere...it might've ended up in my other PR. Essentially, as we're not currently using JSON returns for the DA API, it returns the login page in HTML if certain requests fail. As such, looking for `<!doctype html>` handles this edge case. I'll add a comment above this line!

File: src/library/Server/Manager/Directadmin.php (hunk 5 above)
Comment (wolveix) [reply src/library/Server/Manager/Directadmin.php:83]: Because `is_numeric` can return a float if a float is passed by the user. I've replaced the function with `is_int` instead :)

File: src/library/Server/Manager/Directadmin.php (hunk 7 above)
Comment (wolveix) [reply src/library/Server/Manager/Directadmin.php]: Ah, oversight. Thanks for flagging!

**Comment by wolveix** [comment]
@jaapmarcus @BelleNottelling mentioned this earlier in the PR: 

> 3\. It's not really specific to this PR and is just generally how things are setup right now, but I think we need to stop doing separate DB fields for these configs and instead move to using JSON in a single column. It's more flexible for server managers and would allow us to continue what was started in

I completely agree that that would be a better approach, though given Belle's comment, I figured it'd make more sense to split that off into a new PR. Happy to rework otherwise!

File: src/library/Server/Manager/Directadmin.php (hunk 5 above)
Comment (wolveix) [reply src/library/Server/Manager/Directadmin.php]: Ah, although, `is_int` will fail `'2222'`. Maybe sticking with `is_numeric` makes sense for now

Comment (BelleNottelling) [reply src/library/Server/Manager/Directadmin.php]: Well I was going to be happy with this `is_numeric` explanation, but I won't complain with a more robust implementation.
Thanks!

**Review by BelleNottelling — APPROVED** [review]
All of the smaller items I had noticed have been resolved so I'm pleased.
It needs to get the merge conflicts resolved so tests can run and then once approved you'll have to get the branch up-to-date with our main branch before it can be merged.

Thanks for the quality PR!

**Comment by jaapmarcus** [comment]
> @jaapmarcus @BelleNottelling mentioned this earlier in the PR:
> 
> > 3. It's not really specific to this PR and is just generally how things are setup right now, but I think we need to stop doing separate DB fields for these configs and instead move to using JSON in a single column. It's more flexible for server managers and would allow us to continue what was started in
> 
> I completely agree that that would be a better approach, though given Belle's comment, I figured it'd make more sense to split that off into a new PR. Happy to rework otherwise!

If we going to change the behaviour any way we should do it now. As it also requires database structure updates that need to be altered again with the other PR.  It doesn't make sense to have the changes for maybe a few releases present ...

**Comment by wolveix** [comment]
> If we going to change the behaviour any way we should do it now. As it also requires database structure updates that need to be altered again with the other PR. It doesn't make sense to have the changes for maybe a few releases present ...

As I said, I'm happy to work on this :) Though given my long break from PHP and my freshness to this repo, it may make more sense for @BelleNottelling or someone else to tackle the task. But I'm happy to give it a go otherwise!

**Comment by BelleNottelling** [comment]
> If we going to change the behaviour any way we should do it now. As it also requires database structure updates that need to be altered again with the other PR. It doesn't make sense to have the changes for maybe a few releases present ...

While I understand the point here, I think there's also the argument that @wolveix is following the existing implementation and that it's not their responsibility to completely refactor legacy code for us, most certainly not as one of their first contributions to the code base.

Furthermore, there's an argument that such a change would need to be delegated to at the soonest a 0.7.0 release as that change would be deleting at least 10 columns from the `service_hosting_server` table and would be a very breaking change that any code that directly references any of these columns in the database.

It also needs to be handled and implemented very carefully as any error in the migration step could potentially leave any hosting related service in a at best broken state or at worst accidentally result in essentially a complete lost to the configurations for all of someone's hosting servers within FOSSBIlling.

I really don't think we want to nor should be putting this onto a brand new contributor and I also don't think we should be preventing the merge of this pull request over it. At best, such a refactor is likely months away

**Review by jaapmarcus — APPROVED** [review]
